srtcore/core.cpp

```diff
@@ -8045,6 +7853,207 @@ void CUDT::sendCtrl(UDTMessageType pkttype, const int32_t* lparam, void* rparam,
         m_tsLastSndTime = steady_clock::now();
 }
 
+int CUDT::sendCtrlAck(CPacket& ctrlpkt, int size)
+{
+    SRT_ASSERT(ctrlpkt.getMsgTimeStamp() != 0);
+    int32_t ack;
+    int nbsent = 0;
+    int local_prevack = 0;
+
+#if ENABLE_HEAVY_LOGGING
+    struct SaveBack
+    {
+        int& target;
+        const int& source;
+
+        ~SaveBack() { target = source; }
+    } l_saveback = { m_iDebugPrevLastAck, m_iRcvLastAck };
+    (void)l_saveback; // kill compiler warning: unused variable `l_saveback` [-Wunused-variable]
+
+    local_prevack = m_iDebugPrevLastAck;
+
+    string reason; // just for "a reason" of giving particular % for ACK
+#endif
+
+    // If there is no loss, the ACK is the current largest sequence number plus 1;
+    // Otherwise it is the smallest sequence number in the receiver loss list.
+    if (m_pRcvLossList->getLossLength() == 0)
+    {
+        ack = CSeqNo::incseq(m_iRcvCurrSeqNo);
+#if ENABLE_HEAVY_LOGGING
+        reason = "expected next";
+#endif
+    }
+    else
+    {
+        ScopedLock lock(m_RcvLossLock);
+        ack = m_pRcvLossList->getFirstLostSeq();
+#if ENABLE_HEAVY_LOGGING
+        reason = "first lost";
+#endif
+    }
+
+    if (m_iRcvLastAckAck == ack)
+    {
+        HLOGC(xtlog.Debug, log << "sendCtrl(UMSG_ACK): last ACK %" << ack << "(" << reason << ") == last ACKACK");
+        return nbsent;
+    }
+
+    // send out a lite ACK
+    // to save time on buffer processing and bandwidth/AS measurement, a lite ACK only feeds back an ACK number
+    if (size == SEND_LITE_ACK)
+    {
+        ctrlpkt.pack(UMSG_ACK, NULL, &ack, size);
+        ctrlpkt.m_iID = m_PeerID;
+        nbsent = m_pSndQueue->sendto(m_PeerAddr, ctrlpkt);
+        DebugAck("sendCtrl(lite):" + CONID(), local_prevack, ack);
+        return nbsent;
+    }
+
+    // There are new received packets to acknowledge, update related information.
+    /* tsbpd thread may also call ackData when skipping packet so protect code */
+    enterCS(m_RcvBufferLock);
+
+    // IF ack %> m_iRcvLastAck
+    if (CSeqNo::seqcmp(ack, m_iRcvLastAck) > 0)
+    {
+        const int32_t first_seq ATR_UNUSED = ackDataUpTo(ack);
+        leaveCS(m_RcvBufferLock);
```

Comment:
As this is now part of the function, I think you might consider changing the lock on `m_RcvBuffferLock` to use scope-based tools, `UniqueLock` and `InvertedLock`. Especially if this is refax and aimed at improving readability and error-resistance.

Reply:
Replaced critical section with `UniqueLock`.